File: src/limitedwip/components/ChangeSize.java
package limitedwip.components;

public class ChangeSize {
    public final int value;
    public final boolean isApproximate;

    public ChangeSize(int value) {
        this(value, false);
    }

    public ChangeSize(int value, boolean isApproximate) {
        this.value = value;
        this.isApproximate = isApproximate;
    }

    public ChangeSize add(ChangeSize that) {
        return new ChangeSize(this.value + that.value, this.isApproximate | that.isApproximate);
    }

    @Override public String toString() {
        String result = "ChangeSize(" + value;
        if (isApproximate) result += ", isApproximate";
        return result + ")";
    }

    @Override public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        ChangeSize that = (ChangeSize) o;

        return value == that.value && isApproximate == that.isApproximate;
    }

    @Override public int hashCode() {
        int result = value;
        result = 31 * result + (isApproximate ? 1 : 0);
        return result;
    }
}

File: src/limitedwip/watchdog/ChangeSize.kt
package limitedwip.watchdog

data class ChangeSize(
    val value: Int,
    val isApproximate: Boolean = false
) {
    operator fun plus(that: ChangeSize) = ChangeSize(value + that.value, isApproximate or that.isApproximate)

    companion object {
        val NA = ChangeSize(-1)
        val empty = ChangeSize(0)
        val approximatelyEmpty = ChangeSize(0, isApproximate = true)
    }
}

data class ChangeSizesWithPath(val value: List<Pair<String, ChangeSize>>) {

    val totalChangeSize = value.map { it.second }.fold(ChangeSize.empty) { acc, it -> acc + it }

    fun add(path: String, changeSize: ChangeSize) =
        ChangeSizesWithPath(value + Pair(path, changeSize))

    companion object {
        val empty = ChangeSizesWithPath(emptyList())
    }
}